File: lotto_groups.js
const gerarNumerosDe1a25 = () => Array.from({ length: 25 }, (_, i) => i + 1)

const sortear = (size, numeros) => {
  const sorteados = [];
  while (sorteados.length < size) {
    const index = Math.floor(Math.random() * numeros.length);
    sorteados.push(numeros.splice(index, 1)[0]);
  }
  return sorteados.sort((a, b) => a - b)
}

const sortear15Numeros = () => sortear(15,gerarNumerosDe1a25())

function criarJogosAleatorios() {
  const jogos = [];
  for (let i=0; i<10; i++) jogos.push(sortear15Numeros())
  console.log("10 jogos aleatorios:\n", jogos,"\n");
  return jogos
}

function criarGruposFixos() {
  const grupos = [];
  const numeros = gerarNumerosDe1a25();
  for (let i=0; i<5; i++) grupos[i] = sortear(5, numeros)
  console.log("Grupos de 5 numeros nao-repetidos:\n", grupos,"\n")
  return grupos
}

function combinarGruposEmJogos(grupos) {
  const jogos = [];
  for (let i = 0; i < grupos.length - 2; i++) {
    for (let j = i + 1; j < grupos.length - 1; j++) {
      for (let k = j + 1; k < grupos.length; k++) {
        const jogo = [...grupos[i], ...grupos[j], ...grupos[k]].sort((a, b) => a - b);
        jogos.push(jogo);
      }
    }
  }
  console.log("10 jogos formados pelos grupos:\n", jogos,"\n")
  return jogos
}

// function contarAcertosGrupos(grupos, sorteio) {
//   return grupos
//     .map(grupo => grupo.filter(num => sorteio.includes(num)).length)
//     .sort((a, b) => b - a)
//     .slice(0, 3)
//     .reduce((a, b) => a + b, 0)
// }

const contarAcertos = (jogos, sorteio) => jogos.map(jogo => jogo.filter(num => sorteio.includes(num)).length)

function simular(numSorteios = 100000) {
  const grupos = criarGruposFixos()
  const jogosGrupos = combinarGruposEmJogos(grupos)
  const jogosAleatorios = criarJogosAleatorios()
  const resultados = {grupos: {}, aleatorios: {}}

  for (let i=0; i<numSorteios; i++) {
    const sorteio = sortear15Numeros();

    // const acertosGrupos = contarAcertosGrupos(grupos, sorteio);
    // resultados.grupos[acertosGrupos] = (resultados.grupos[acertosGrupos] || 0) + 1;

    const acertosGrupos = contarAcertos(jogosGrupos, sorteio);
    acertosGrupos.forEach(acertos => {
      resultados.grupos[acertos] = (resultados.grupos[acertos] || 0) + 1;
    });

    const acertosAleatorios = contarAcertos(jogosAleatorios, sorteio);
    acertosAleatorios.forEach(acertos => {
      resultados.aleatorios[acertos] = (resultados.aleatorios[acertos] || 0) + 1;
    });
  }
  return resultados;
}

const percentage = (resultado, total) => ((resultado || 0) / total * 100)
const print = (e, d = 2, n = 10) =>  e.toFixed(d).padStart(n)

function mostrarEstatisticas(resultados, totalSorteios) {
  let gruposAcumulado = 0;
  let aleatoriosAcumulado = 0;

  console.log("\n Acertos |   Grupos (%)  | Aleatórios (%)");
  for (let i = 15; i >= 5; i--) {
    const g = percentage(resultados.grupos[i],totalSorteios * 10);
    const a = percentage(resultados.aleatorios[i],totalSorteios * 10);
    gruposAcumulado += g
    aleatoriosAcumulado += a

    console.log(`   ${print(i,0,3)}   |${print(g)}     |${print(a)}`);
  }
  console.log(`   Total |${print(gruposAcumulado)}     |${print(aleatoriosAcumulado)}`);

  console.log(`\nNumero de sorteios simulados: ${totalSorteios.toLocaleString("pt-BR")}`);
  console.log("------------------------");
  console.log("");
}

// Executar
const totalSimulacoes = 3268760*2;
const resultados = simular(totalSimulacoes,);
mostrarEstatisticas(resultados, totalSimulacoes);

File: lotto_groups.py
import random
from collections import defaultdict

def gerar_numeros_de_1_a_25():
    return list(range(1, 26))

def sortear(size, numeros_originais):
    numeros = numeros_originais.copy()
    sorteados = []
    while len(sorteados) < size:
        index = random.randrange(len(numeros))
        sorteados.append(numeros.pop(index))
    return sorted(sorteados)

def sortear_15_numeros():
    return sortear(15, gerar_numeros_de_1_a_25())

def criar_jogos_aleatorios():
    jogos = [sortear_15_numeros() for _ in range(10)]
    print("10 jogos aleatórios:\n", jogos, "\n")
    return jogos

def criar_grupos_fixos():
    numeros = gerar_numeros_de_1_a_25()
    grupos = [sortear(5, numeros) for _ in range(5)]
    print("Grupos de 5 números não-repetidos:\n", grupos, "\n")
    return grupos

def combinar_grupos_em_jogos(grupos):
    jogos = []
    for i in range(len(grupos) - 2):
        for j in range(i + 1, len(grupos) - 1):
            for k in range(j + 1, len(grupos)):
                jogo = sorted(grupos[i] + grupos[j] + grupos[k])
                jogos.append(jogo)
    print("10 jogos formados pelos grupos:\n", jogos, "\n")
    return jogos

def contar_acertos(jogos, sorteio):
    return [len(set(jogo) & set(sorteio)) for jogo in jogos]

def simular(num_sorteios=100_000):
    grupos = criar_grupos_fixos()
    jogos_grupos = combinar_grupos_em_jogos(grupos)
    jogos_aleatorios = criar_jogos_aleatorios()
    resultados = {'grupos': defaultdict(int), 'aleatorios': defaultdict(int)}

    for _ in range(num_sorteios):
        sorteio = sortear_15_numeros()

        acertos_grupos = contar_acertos(jogos_grupos, sorteio)
        for acertos in acertos_grupos:
            resultados['grupos'][acertos] += 1

        acertos_aleatorios = contar_acertos(jogos_aleatorios, sorteio)
        for acertos in acertos_aleatorios:
            resultados['aleatorios'][acertos] += 1

    return resultados

def percentage(resultado, total):
    return (resultado / total * 100) if resultado else 0

def print_alinhado(valor, d=2, n=10):
    return f"{valor:.{d}f}".rjust(n)

def mostrar_estatisticas(resultados, total_sorteios):
    total_jogos = total_sorteios * 10
    grupos_acumulado = 0
    aleatorios_acumulado = 0

    print("\n Acertos |   Grupos (%)  | Aleatórios (%)")
    for i in range(15, 4, -1):
        g = percentage(resultados['grupos'].get(i, 0), total_jogos)
        a = percentage(resultados['aleatorios'].get(i, 0), total_jogos)
        grupos_acumulado += g
        aleatorios_acumulado += a
        print(f"   {str(i).rjust(3)}   |{print_alinhado(g)}     |{print_alinhado(a)}")

    print(f"   Total |{print_alinhado(grupos_acumulado)}     |{print_alinhado(aleatorios_acumulado)}")
    print(f"\nNúmero de sorteios simulados: {total_sorteios:,}".replace(",", "."))
    print("------------------------\n")

# Executar
if __name__ == "__main__":
    total_simulacoes = 3_268_760 * 2
    resultados = simular(total_simulacoes)
    mostrar_estatisticas(resultados, total_simulacoes)
